Device tracker: SNMPv3

T3m3z opened the pull request:
**Description:**
Original SNMP device tracker only supported SNMPv1 where security is non-existent. This PR adds support for SNMPv3 with authorization using SHA and privacy using AES. 

Functionality was tested using Mikrotik router. Component tries to use SNMPv3 if both authkey (key for authentication) and privkey (key for encrypting data) are provided. Otherwise it will revert back to older SNMP version and use only community-string.

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
device_tracker:
  - platform: snmp
    host: 192.168.0.1
    community: "community"
    authkey: "pass1234"
    privkey: "pass1234"
    baseoid: 1.3.6.1.2.1.4.22.1.2
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [x] New files were added to `.coveragerc`.

T3m3z committed:
Add initial SNMPv3 support for better security

T3m3z committed:
Fixed indentation errors

T3m3z committed:
Fixed flake8 E128 on row 65

T3m3z committed:
Disabled warning about too many instance-attributes

kellerza commented on homeassistant/components/device_tracker/snmp.py:
```diff
@@ -94,8 +113,12 @@ def get_snmp_data(self):
 
-        errindication, errstatus, errindex, restable = self.snmp.nextCmd(
-            self.community, self.host, self.baseoid)
+        if self.version == 3:
```
If you use the same variable above, say `self.request`, instead of `self.community` and `self.userdata` then you don't require this if statement. You also don't need the `self.version`

T3m3z committed:
Removed extra code, added Inclusive to make better config validation

T3m3z commented:
Thanks @kellerza ! I added that `vol.Inclusive` to PR and removed those unnecessary parts of the code. That `self.version` part was stupid brainfart :)

kellerza commented:
Looks good! :dolphin: 

Will you please create a PR on the snmp docs?

T3m3z commented:
Of course! I don't have time today/tomorrow but probably during weekend I can make it.
